imap/autocreate.c

```diff
@@ -93,330 +93,398 @@ static const char *get_script_name(const char *filename)
-    if(!(compiled_source_script = config_getstring(IMAPOPT_AUTOCREATE_SIEVE_SCRIPT_COMPILED))) {
-        syslog(LOG_WARNING, "autocreate_sieve: autocreate_sieve_compiledscript option is not defined. Compiling it");
+    compiled_source_script = config_getstring(IMAPOPT_AUTOCREATE_SIEVE_SCRIPT_COMPILED);
+    if(!compiled_source_script) {
+        syslog(LOG_WARNING, "autocreate_sieve: autocreate_sieve_compiledscript"
+               "option is not defined. Compiling it");
         do_compile = 1;
     }
 
-    if (!(sieve_script_dir = user_sieve_path(userid))) {
-        syslog(LOG_WARNING, "autocreate_sieve: unable to determine sieve directory for user %s", userid);
-        return 1;
+    sieve_script_dir = user_sieve_path(userid);
+    if (!sieve_script_dir) {
+        syslog(LOG_ERR, "autocreate_sieve: unable to determine sieve directory"
+               "for user %s", userid);
+        goto failed_start;
     }
 
-    if(snprintf(sieve_tmpname, MAX_FILENAME, "%s/%s.script.NEW",sieve_script_dir, sievename) >= MAX_FILENAME) {
-        syslog(LOG_WARNING, "autocreate_sieve: Invalid sieve path %s, %s, %s", sieve_script_dir, sievename, userid);
-        return 1;
-    }
-    if(snprintf(sieve_bctmpname, MAX_FILENAME, "%s/%s.bc.NEW",sieve_script_dir, sievename) >= MAX_FILENAME) {
-        syslog(LOG_WARNING, "autocreate_sieve: Invalid sieve path %s, %s, %s", sieve_script_dir, sievename, userid);
-        return 1;
-    }
-    if(snprintf(sieve_script_name, MAX_FILENAME, "%s/%s.script",sieve_script_dir, sievename) >= MAX_FILENAME) {
-        syslog(LOG_WARNING, "autocreate_sieve: Invalid sieve path %s, %s, %s", sieve_script_dir, sievename, userid);
-        return 1;
-    }
-    if(snprintf(sieve_bcscript_name, MAX_FILENAME, "%s/%s.bc",sieve_script_dir, sievename) >= MAX_FILENAME) {
-        syslog(LOG_WARNING, "autocreate_sieve: Invalid sieve path %s, %s, %s", sieve_script_dir, sievename, userid);
-        return 1;
-    }
-    if(snprintf(sieve_default, MAX_FILENAME, "%s/%s",sieve_script_dir,"defaultbc") >= MAX_FILENAME) {
-        syslog(LOG_WARNING, "autocreate_sieve: Invalid sieve path %s, %s, %s", sieve_script_dir, sievename, userid);
-        return 1;
-    }
-    /* XXX no directory? umm */
-    if(snprintf(sieve_bclink_name, MAX_FILENAME, "%s.bc", sievename) >= MAX_FILENAME) {
-        syslog(LOG_WARNING, "autocreate_sieve: Invalid sieve path %s, %s, %s", sieve_script_dir, sievename, userid);
-        return 1;
+    if (setup_sieve_filenames(sieve_script_dir, sievename, &script_names) != 0) {
+        syslog(LOG_ERR, "autocreate_sieve: Invalid sieve path %s, %s, %s",
+               sieve_script_dir, sievename, userid);
+        goto failed_start;
     }
 
     /* Check if a default sieve filter already exists */
-    if(!stat(sieve_default,&statbuf)) {
-        syslog(LOG_WARNING,"autocreate_sieve: Default sieve script already exists");
-        return 1;
+    if(!stat(script_names.defaultname, &statbuf)) {
```

Comment:
ack! I missed a few of these. :-/ I wanted to rewrite the whole thing, but it gave me a headache trying to figure out the code flow, so thought I would first bring some semblance of sanity and go from there.